replit_river/client_transport.py

```diff
@@ -104,12 +104,25 @@ async def _establish_new_connection(
         client_id = self._client_id
         logger.info("Attempting to establish new ws connection")
 
+        last_error = None
         for i in range(max_retry):
             if i > 0:
                 logger.info(f"Retrying build handshake number {i} times")
             if not rate_limit.has_budget(client_id):
                 logger.debug("No retry budget for %s.", client_id)
-                break
+                if last_error:
+                    raise RiverException(
+                        ERROR_HANDSHAKE,
+                        (
+                            f"No retry budget for {client_id} after "
+                            f"last failure: {str(last_error)}"
+                        ),
+                    )
```

Comment:
```suggestion
                    raise RiverException(
                        ERROR_HANDSHAKE,
                        f"No retry budget for {client_id}"
                    ) from last_error
```

it's better to rely on Python to do the concatenation for us. that way we also preserve the OG stack :massage: